src/backend/distributed/master/master_citus_tools.c

```diff
@@ -0,0 +1,538 @@
+ * to connect to remote nodes.
+ */
+Datum
+master_run_on_worker(PG_FUNCTION_ARGS)
+{
+	ReturnSetInfo *rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
+	MemoryContext per_query_ctx = NULL;
+	MemoryContext oldcontext = NULL;
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = NULL;
+	AttInMetadata *attinmeta = NULL;
+	bool parallelExecution = false;
+	StringInfo *nodeNameArray = NULL;
+	int *nodePortArray = NULL;
+	StringInfo *commandStringArray = NULL;
+	bool *statusArray = NULL;
+	StringInfo *resultArray = NULL;
+	int commandIndex = 0;
+	int commandCount = 0;
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (!rsinfo || !(rsinfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_SYNTAX_ERROR),
+				 errmsg("materialize mode required, but it is not "
+						"allowed in this context")));
+	}
+
+	commandCount = ParseCommandParameters(fcinfo, &nodeNameArray, &nodePortArray,
+										  &commandStringArray, &parallelExecution);
+
+	per_query_ctx = rsinfo->econtext->ecxt_per_query_memory;
+	oldcontext = MemoryContextSwitchTo(per_query_ctx);
+
+	/* get the requested return tuple description */
+	tupleDescriptor = CreateTupleDescCopy(rsinfo->expectedDesc);
+
+	/*
+	 * Check to make sure we have correct tuple descriptor
+	 */
+	if (tupleDescriptor->natts != 4 ||
+		tupleDescriptor->attrs[0]->atttypid != TEXTOID ||
+		tupleDescriptor->attrs[1]->atttypid != INT4OID ||
+		tupleDescriptor->attrs[2]->atttypid != BOOLOID ||
+		tupleDescriptor->attrs[3]->atttypid != TEXTOID)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_COLUMN_DEFINITION),
+				 errmsg("query-specified return tuple and "
+						"function return type are not compatible")));
+	}
+
+	/* prepare storage for status and result values */
+	statusArray = palloc0(commandCount * sizeof(bool));
+	resultArray = palloc0(commandCount * sizeof(StringInfo));
+	for (commandIndex = 0; commandIndex < commandCount; commandIndex++)
+	{
+		resultArray[commandIndex] = makeStringInfo();
+	}
+
+	if (parallelExecution)
+	{
+		ExecuteCommandsInParallelAndStoreResults(nodeNameArray, nodePortArray,
+												 commandStringArray,
+												 statusArray, resultArray, commandCount);
+	}
+	else
+	{
+		ExecuteCommandsAndStoreResults(nodeNameArray, nodePortArray, commandStringArray,
+									   statusArray, resultArray, commandCount);
+	}
+
+	/* let the caller know we're sending back a tuplestore */
+	rsinfo->returnMode = SFRM_Materialize;
+	tupleStore = tuplestore_begin_heap(true, false, work_mem);
+	attinmeta = TupleDescGetAttInMetadata(tupleDescriptor);
+
+	for (commandIndex = 0; commandIndex < commandCount; commandIndex++)
+	{
+		HeapTuple tuple = NULL;
+		char *values[4];
+
+		values[0] = nodeNameArray[commandIndex]->data;
+		values[1] = palloc0(12); /* enough to hold an integer */
+		pg_ltoa(nodePortArray[commandIndex], values[1]);
+		values[2] = statusArray[commandIndex] ? "t" : "f";
+		values[3] = resultArray[commandIndex]->data;
```

Comment:
I really would recommend rather using heap_form_tuple and values/isnull arrays.  I don't think it's a blocker, but imo it's clearly better than using BuildTupleFromCStrings and doing integer -> text -> integer conversions.